src/VisualStudio/CSharp/Impl/CSharpPackage.cs

```diff
@@ -93,15 +96,20 @@ protected override VisualStudioWorkspaceImpl CreateWorkspace()
 
         protected override void Dispose(bool disposing)
         {
-            UnregisterObjectBrowserLibraryManager();
+            if (disposing)
+            {
+                JoinableTaskFactory.Run(() => UnregisterObjectBrowserLibraryManagerAsync(CancellationToken.None));
+            }
 
             base.Dispose(disposing);
         }
 
-        private void RegisterObjectBrowserLibraryManager()
+        private async Task RegisterObjectBrowserLibraryManagerAsync(CancellationToken cancellationToken)
         {
-            if (this.GetService(typeof(SVsObjectManager)) is IVsObjectManager2 objectManager)
+            if (await GetServiceAsync(typeof(SVsObjectManager)).ConfigureAwait(true) is IVsObjectManager2 objectManager)
             {
+                await JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
```

Comment:
Switch back to main thread here seems odd given we already started on the main thread when this was called from InitializeAsync().

Reply:
This is for code resiliency. Per the 1st threading rule, all methods should take ownership for the thread they run on. For async methods in particular, they are supposed to be safe to call from any thread (since they can switch to the thread they need). So we shouldn't skip this call based on the *current* codebase only calling it from a UI thread context, as that makes for fragile code and weakens people's confidence in the safety the threading rules provide.
When you're already on the main thread, this switch call is a fast no-op.

Reply:
Yep, defensiviness is correct. My concern is this is indistuingishable between "we meant to do that" and "we copied code around and this ended up this way". In some places in Roslyn we have a "I can be called on any thread" fake method that we use to mark the intent; do we have a equivalent we can use here?

Reply:
The 1st threading rule states that *all* async methods have the 'I can be called on any thread' status. So no need for a no-op method call to indicate that. (if you're following the rules).
For sync methods, we don't have an equivalent at the VS SDK level. We do have the `ThreadHelper.ThrowIfNotOnUIThread` method, which with our latest threading analyzers propagates to help you add that to other methods that call it. So if you're leveraging the analyzers, you can come to trust the *lack* of such a method call at the top of your method body as a pretty good guarantee that it's free-threaded.

Reply:
IVsObjectManager seems STA service, is GetServiceAsync okay to be used in BG for such service?

Reply:
> IVsObjectManager seems STA service, is GetServiceAsync okay to be used in BG for such service?

`GetServiceAsync` is fine, but the `is IVsObjectManager2` cast needs to occur on the main thread. I've updated the code in 7b7c87da83e465ed489a1aa5de856bb3fca701d3 to correct this.